File: clj/src/slipstream/ui/views/subsection.clj
(ns slipstream.ui.views.subsection
  (:require [net.cgrand.enlive-html :as html]
            [slipstream.ui.util.core :as u]
            [slipstream.ui.util.enlive :as ue]
            [slipstream.ui.util.clojure :as uc]))

(def template-filename (u/template-path-for "subsection.html"))

(def ^:private subsection-selected-cls "active")

(def ^:private activator-group-sel    [:.ss-subsection-activator-group])
(def ^:private activator-sel          [[:li html/first-of-type]])
(def ^:private activator-xs-group-sel [:.ss-subsection-activator-xs-group])
(def ^:private activator-xs-sel       [[:option html/first-of-type]])
(def ^:private content-group-sel      [:.ss-subsection-content-group])
(def ^:private content-sel            [[:div.tab-pane html/first-of-type]])

(defn- activator-group
  [subsections]
  (ue/content-for activator-sel [{:keys [title selected? id]} subsections]
                  ue/this  (ue/enable-class selected? subsection-selected-cls)
                  [:a]    (ue/set-href "#" id)
                  [:a]    (html/content (str title))))

(defn- activator-xs-group
  [subsections]
  (ue/content-for activator-xs-sel [{:keys [title id]} subsections]
                  ue/this  (ue/set-value "#" id)
                  ue/this  (html/content (str title))))

(defn- content-group
  [subsections]
  (ue/content-for content-sel [{:keys [content selected? id]} subsections]
                  ue/this  (ue/enable-class selected? subsection-selected-cls)
                  ue/this  (ue/set-id id)
                  ue/this  (if (string? content)
                            (html/html-content content)
                            (html/content content))))

;; TODO: Refactor the 'subsection-group-snip-*' into one, since the body is identical.

(html/defsnippet subsection-group-snip-flat template-filename [:.ss-subsection-group]
  [subsections]
  activator-group-sel     (activator-group subsections)
  activator-xs-group-sel  (activator-xs-group subsections)
  content-group-sel       (content-group subsections))

(html/defsnippet subsection-group-snip-stacked template-filename [:.ss-subsection-group-stacked]
  [subsections]
  activator-group-sel     (activator-group subsections)
  activator-xs-group-sel  (activator-xs-group subsections)
  content-group-sel       (content-group subsections))

(def ^:private max-num-of-flat-sections 7)

(defn- subsection-group-snip
  [subsections]
  (if (-> subsections count (> max-num-of-flat-sections))
    (subsection-group-snip-stacked subsections)
    (subsection-group-snip-flat subsections)))

(defn- assoc-id
  [subsection-id-prefix subsection]
  (assoc subsection :id (->> subsection :title uc/keywordize name (str subsection-id-prefix "-"))))

(defn build
  [subsections]
  (html/content (subsection-group-snip (->> subsections
                                            vec
                                            u/ensure-one-selected
                                            (map (partial assoc-id (gensym "ss-subsection-")))))))

File: clj/resources/static_content/js/subsection.js
jQuery( function() { ( function( $$, $, undefined ) {

    function selectFromCombobox(e) {
        var $combobox = $(e),
            targetTab = $combobox.children("option:selected").attr("value");
            $tabAnchor = $combobox.parent().find("a[href='" + targetTab + "']");
        $tabAnchor.tab("show");
        return;
    }

    function enableSubsections() {
        // Match the height of the tab contents to the highest to avoid jumping
        // NOTE: Doesn't work if the element is not visible on page load (e.g.
        //       in a hidden section) since the height will be 0.
        $(".ss-subsection-group, .ss-subsection-group-stacked, .ss-subsection-group-dropdown").each(function (){
            var $thisTabPanes = $(this).find(".tab-pane"),
                highestHight = Math.max.apply(null,
                    $thisTabPanes.map(function(){
                        return $(this).height();
                    }));
            $thisTabPanes.css("min-height", highestHight);
            return;
        });

        // Sync the combobox .ss-subsection-activator-xs-group at page load
        $(".ss-subsection-activator-xs-group").each(function (){
            var $this = $(this),
                activeSubsectionId = $this.parent().find("li.active a").attr("href");
            $this.val(activeSubsectionId);
        });

        // Enable tab functionality, keeping in sync the combobox .ss-subsection-activator-xs-group
        $(".ss-subsection-activator-group a").click(function (e) {
            var $tabAnchor  = $(this),
                targetTab   = $tabAnchor.attr("href"),
                $combobox   = $tabAnchor.closest("div").find(".ss-subsection-activator-xs-group");
            e.preventDefault();
            $combobox.val(targetTab);
            $tabAnchor.tab("show");
        });

        // Enable ss-subsection-activator-xs-group
        $(".ss-subsection-activator-xs-group").change( function () {
            selectFromCombobox(this);
        });
    }

    enableSubsections();

    var subsectionIdPrefix = "ss-subsection-",
        onShowCallbackKey = "on-show-callback",
        tabAnchorSel = ".ss-subsection-activator-group a[role=tab]";

    $(tabAnchorSel).on("shown.bs.tab", function (e, extraParametersArg) {

        // Ensure correct hash when opening subsections
        var $tabAnchor          = $(this),
            subsectionTitle     = $tabAnchor.text(),
            subsectionId        = $tabAnchor.attr("href"),
            subsectionIdTrimmed = subsectionId
                                    .trimPrefix("#" + subsectionIdPrefix)
                                    .replace(/\d+-/, ""), // remove digits of the unique prefix
            $subsectionContent  = $(subsectionId),
            onShowCallback      = $tabAnchor.data(onShowCallbackKey),
            extraParameters = $.extend({
                    // In flat sections, do not persist the open subsection in the hash
                    skipHashUpdate: $tabAnchor.closest(".ss-section-flat").foundAny()
                },
                extraParametersArg);

        if (! extraParameters.skipHashUpdate) {
            $$.util.url.hash.updateValues({1: subsectionIdTrimmed});
        }

        // Run on-show-callback if present
        if ($.isFunction(onShowCallback)) {
            onShowCallback(subsectionTitle, $subsectionContent);
        }
    });

    $$.subsection = {
        reenableSubsections: enableSubsections,

        selectByTitle: function($section, title) {
            return $section
                       .find(".ss-subsection-activator-group")
                           .find("a[href^='#" + subsectionIdPrefix + "'][href$='" + title.toLowerCase() + "']")
                               .click()
                               .foundOne();
        },

        // Callback with the signature: function(subsectionTitle, $subsectionContent)
        onShow: function(callback) {
            $(tabAnchorSel).data(onShowCallbackKey, callback);
        },
        triggerOnShowOnOpenSubsection: function(skipHashUpdate) {
            var extraParameters = {
                skipHashUpdate: skipHashUpdate
            };
            $(".panel .panel-collapse.collapse.in, .ss-section-flat")       // The open collapsible section and the flat sections
                .find(".ss-subsection-group, .ss-subsection-group-stacked") // Its subsections (horizontal, vertical or combobox)
                    .find("li.active > a")                                  // The open subsection
                        .trigger("show.bs.tab", extraParameters)
                        .trigger("shown.bs.tab", extraParameters);
       }

    };

}( window.SlipStream = window.SlipStream || {}, jQuery ));});
